File: 2117-find-original-array-from-doubled-array/find-original-array-from-doubled-array.py
# Optimal Code
class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        # total number of elements
        n = len(changed)
        # valid doubled array must have an even number of elements
        if n % 2 != 0:
            return []
        # sort the array so that smaller number are processed first
        changed.sort()
        # dictionary to map : number -> frequency
        mp = {}
        # count the frequency of evey number
        for num in changed:
            mp[num] = mp.get(num, 0)  + 1
        # store the reconstruted original array
        result = []
        # process the number in sorted orders
        for num in changed:
            # double of the current number
            twice = 2 * num
            # skip number that have already been paired
            if mp[num] == 0:
                continue
            # if the double does not exist or has already been used the array is invalid
            if twice not in mp or mp[twice] == 0:
                return []
            # add the current number to the original array
            result.append(num)
            # use one occurence of the number and its double
            mp[num] -= 1
            mp[twice] -= 1
        # return the recoverd original array
        return result 

# Time Complexity : O(Nlog)
# Space Complexity : O(N)
        

File: 2117-find-original-array-from-doubled-array/find-original-array-from-doubled-array.cpp
// Optimal Code
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        // total number of elements
        int n = changed.size();
        // valid doubled array must have an even number of elements
        if (n % 2 != 0) {
            return {};
        }
        // sort the array so that smaller numbers are processed first
        sort(changed.begin(), changed.end());
        // dictionary to map: number -> frequency
        unordered_map<int, int> mp;
        // count the frequency of every number
        for (int num : changed) {
            mp[num]++;
        }
        // store the reconstructed original array
        vector<int> result;
        // process the numbers in sorted order
        for (int num : changed) {
            // skip numbers that have already been paired
            if (mp[num] == 0) {
                continue;
            }
            // double of the current number
            int twice = 2 * num;
            // if the double does not exist or has already been used the array is invalid
            if (mp.find(twice) == mp.end() || mp[twice] == 0) {
                return {};
            }
            // add the current number to the original array
            result.push_back(num);
            // use one occurrence of the number and one occurrence of its double
            mp[num]--;
            mp[twice]--;
        }
        // return the recovered original array
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)